File: src/forager/library/scanner.py
from __future__ import annotations
import re
from pathlib import Path
from forager.core.game import ENGINE_NAMES, Game, Source
from forager.core.paths import games_dir

# Steam app IDs that are tools/runtimes rather than games (never shown in
# the library). Proton Experimental is the big one; the Linux runtime
# containers and SteamVR are included so they can't sneak in either.
STEAM_TOOL_APP_IDS = {
    "1493710",  # Proton Experimental
    "250820",   # SteamVR
    "1070560",  # Steam Linux Runtime
    "1391110",  # Steam Linux Runtime - Soldier
    "1628350",  # Steam Linux Runtime - Sniper
}


def scan_all() -> list[Game]:
    seen: set[Game] = set()
    for scanner in (_scan_steam, _scan_minecraft, _scan_standalone):
        for game in scanner():
            if game not in seen:
                seen.add(game)
    return sorted(
        (g for g in seen if "proton" not in g.name.lower()),
        key=lambda g: g.sort_key or g.name.lower(),
    )


def _scan_steam() -> list[Game]:
    games: list[Game] = []
    apps_dir = games_dir() / "steam/steamapps"
    if not apps_dir.is_dir():
        return games

    for acf in sorted(apps_dir.glob("appmanifest_*.acf")):
        app_id, name = _parse_acf(acf)
        if app_id in STEAM_TOOL_APP_IDS:
            continue
        if app_id and name:
            games.append(
                Game(
                    name=name,
                    source=Source.STEAM,
                    path=apps_dir / "common" / name,
                    app_id=app_id,
                    sort_key=name.lower(),
                )
            )
    return games


def _parse_acf(path: Path) -> tuple[str | None, str | None]:
    try:
        text = path.read_text("utf-8", errors="replace")
        app_id = _acf_val(text, "appid")
        name = _acf_val(text, "name")
        if name:
            name = name.removesuffix("\u0000")
        return (app_id, name)
    except Exception:
        return (None, None)


def _acf_val(text: str, key: str) -> str | None:
    m = re.search(rf'"{re.escape(key)}"\s+"(.+?)"', text)
    if m:
        return m.group(1)
    return None


def _scan_minecraft() -> list[Game]:
    games: list[Game] = []
    mc_dir = games_dir() / "minecraft"
    if not mc_dir.is_dir():
        return games

    for entry in sorted(mc_dir.iterdir()):
        if not entry.is_dir():
            continue
        if entry.name.startswith(".") or entry.name == ".LAUNCHER_TEMP":
            continue
        games.append(
            Game(
                name=entry.name,
                source=Source.MINECRAFT,
                path=entry,
                sort_key=entry.name.lower(),
            )
        )
    return games


def _scan_standalone() -> list[Game]:
    games: list[Game] = []
    for container in ("standalone", "drm-free"):
        base = games_dir() / container
        if not base.is_dir():
            continue
        for entry in sorted(base.iterdir()):
            if not entry.is_dir() or entry.name.startswith("."):
                continue
            if entry.name == "series":
                games.extend(_scan_series_dir(entry))
            else:
                games.extend(_scan_loose_root(entry))
    return games


def _scan_loose_root(root: Path) -> list[Game]:
    """Games directly under root, possibly grouped under an engine folder.

    ``drm-free/standalone/other/bdcc`` -> game ``bdcc``.
    """
    games: list[Game] = []
    for entry in sorted(root.iterdir()):
        if not entry.is_dir() or entry.name.startswith("."):
            continue
        if _is_game_dir(entry):
            games.append(_loose_game(entry))
        else:
            games.extend(_scan_loose_flat(entry))
    return games


def _scan_loose_flat(dir: Path) -> list[Game]:
    games: list[Game] = []
    for entry in sorted(dir.iterdir()):
        if not entry.is_dir() or entry.name.startswith("."):
            continue
        games.append(_loose_game(entry))
    return games


def _scan_series_dir(series_root: Path) -> list[Game]:
    """Series games, with an optional engine level that is stripped from names.

    ``series/rpgMaker/sequel/asylum`` -> game ``sequel/asylum``.
    ``series/unity/furry shades of gay/2`` -> game ``furry shades of gay/2``.
    """
    games: list[Game] = []
    for entry in sorted(series_root.iterdir()):
        if not entry.is_dir() or entry.name.startswith("."):
            continue
        if _is_game_dir(entry):
            games.append(_loose_game(entry))
            continue
        parts = [] if entry.name.lower() in ENGINE_NAMES else [entry.name]
        _collect_series(entry, parts, games)
    return games


def _is_game_dir(path: Path) -> bool:
    if (path / "Game.ini").is_file():
        return True
    for pattern in ("*.exe", "*.x86_64", "*.sh", "*.py", "icon.png"):
        if next(path.glob(pattern), None) is not None:
            return True
    return False


def _collect_series(path: Path, parts: list[str], out: list[Game]) -> None:
    if _is_game_dir(path):
        rel = "/".join(parts)
        out.append(
            Game(
                name=rel,
                source=Source.STANDALONE,
                path=path,
                sort_key=rel,
            )
        )
        return
    for entry in sorted(path.iterdir()):
        if entry.is_dir() and not entry.name.startswith("."):
            _collect_series(entry, parts + [entry.name], out)


def _loose_game(entry: Path) -> Game:
    kwargs = dict(
        name=entry.name,
        source=Source.STANDALONE,
        path=entry,
        sort_key=entry.name.lower(),
    )
    if entry.name == "bdcc":
        kwargs["search_names"] = ["Broken Dreams Correctional Center"]
    return Game(**kwargs)

File: src/library/Scanner.cpp
#include "library/Scanner.h"
#include "library/Metadata.h"
#include "core/Paths.h"
#include "utils/Acf.h"

#include <QDir>
#include <QFileInfo>
#include <QSet>
#include <algorithm>

static const QSet<QString> STEAM_TOOL_APP_IDS = {
    "1493710", "250820", "1070560", "1391110", "1628350"
};

static const QSet<QString> GENERIC_CONTAINERS = {
    "standalone", "series", "minecraft", "steam", "rpgmaker", "rpg",
    "games", "instances", "launcher", "single", "flash", "drm-free",
};

static const QSet<QString> ENGINE_NAMES = {"other", "rpgmaker", "unity", "unreal"};

template <typename Fn>
static void forEachSubdir(const QString& path, Fn fn)
{
    QDir dir(path);
    dir.setFilter(QDir::Dirs | QDir::NoDotAndDotDot);
    dir.setSorting(QDir::Name);
    for (const auto& entry : dir.entryInfoList()) {
        if (entry.fileName().startsWith('.')) continue;
        fn(entry);
    }
}

// --- Steam scanner ---

static std::vector<Game> scanSteam()
{
    std::vector<Game> games;
    for (const QString& file : acf::listManifests(paths::gamesDir() + "/steam/steamapps")) {
        auto [appId, name] = acf::parseFile(file);
        if (appId.isEmpty() || name.isEmpty()) continue;
        if (STEAM_TOOL_APP_IDS.contains(appId)) continue;

        Game g(name, Source::Steam);
        g.setAppId(appId);
        g.setPath(QFileInfo(file).absolutePath() + "/common/" + name);
        g.setSortKey(name.toLower());
        games.push_back(std::move(g));
    }
    return games;
}

// --- Minecraft scanner ---

static std::vector<Game> scanMinecraft()
{
    std::vector<Game> games;
    QString root = paths::gamesDir() + "/minecraft";
    if (!QDir(root).exists()) return games;

    forEachSubdir(root, [&](const QFileInfo& entry) {
        QString name = entry.fileName();
        Game g(name, Source::Minecraft);
        g.setPath(entry.absoluteFilePath());
        g.setSortKey(name.toLower());
        games.push_back(std::move(g));
    });
    return games;
}

// --- Standalone scanner ---

static bool isGameDir(const QString& path)
{
    QDir dir(path);
    if (QFileInfo::exists(path + "/Game.ini")) return true;

    for (const auto& pattern : {"*.exe", "*.x86_64", "*.sh", "*.py", "icon.png"}) {
        if (!dir.entryList({pattern}, QDir::Files).isEmpty()) return true;
    }
    return false;
}

static Game makeLooseGame(const QFileInfo& entry)
{
    QString name = entry.fileName();
    Game g(name, Source::Standalone);
    g.setPath(entry.absoluteFilePath());
    g.setSortKey(name.toLower());

    if (name == "bdcc") {
        g.setSearchNames({"Broken Dreams Correctional Center"});
    }
    return g;
}

static void collectSeries(const QString& path, QStringList parts, std::vector<Game>& out);

static void scanLooseFlat(const QString& dirPath, std::vector<Game>& out)
{
    forEachSubdir(dirPath, [&](const QFileInfo& entry) {
        out.push_back(makeLooseGame(entry));
    });
}

static void scanLooseRoot(const QString& rootPath, std::vector<Game>& out)
{
    forEachSubdir(rootPath, [&](const QFileInfo& entry) {
        if (isGameDir(entry.absoluteFilePath())) {
            out.push_back(makeLooseGame(entry));
        } else {
            scanLooseFlat(entry.absoluteFilePath(), out);
        }
    });
}

static void collectSeries(const QString& path, QStringList parts, std::vector<Game>& out)
{
    if (isGameDir(path)) {
        QString rel = parts.join('/');
        Game g(rel, Source::Standalone);
        g.setPath(path);
        g.setSortKey(rel);
        out.push_back(std::move(g));
        return;
    }

    forEachSubdir(path, [&](const QFileInfo& entry) {
        QStringList newParts = parts;
        if (!ENGINE_NAMES.contains(entry.fileName().toLower())) {
            newParts.append(entry.fileName());
        }
        collectSeries(entry.absoluteFilePath(), newParts, out);
    });
}

static void scanSeriesDir(const QString& seriesRoot, std::vector<Game>& out)
{
    forEachSubdir(seriesRoot, [&](const QFileInfo& entry) {
        if (isGameDir(entry.absoluteFilePath())) {
            out.push_back(makeLooseGame(entry));
            return;
        }
        QStringList parts;
        if (!ENGINE_NAMES.contains(entry.fileName().toLower())) {
            parts.append(entry.fileName());
        }
        collectSeries(entry.absoluteFilePath(), parts, out);
    });
}

static std::vector<Game> scanStandalone()
{
    std::vector<Game> games;
    for (const auto& container : {"standalone", "drm-free"}) {
        QString base = paths::gamesDir() + "/" + container;
        if (!QDir(base).exists()) continue;

        forEachSubdir(base, [&](const QFileInfo& entry) {
            if (entry.fileName() == "series") {
                scanSeriesDir(entry.absoluteFilePath(), games);
            } else {
                scanLooseRoot(entry.absoluteFilePath(), games);
            }
        });
    }
    return games;
}

// --- Main entry ---

std::vector<Game> scanner::scanAll()
{
    QSet<QString> seen;
    std::vector<Game> all;

    auto merge = [&](std::vector<Game>&& games) {
        for (auto& g : games) {
            QString id = g.identifier();
            if (seen.contains(id)) continue;
            seen.insert(id);
            all.push_back(std::move(g));
        }
    };

    merge(scanSteam());
    merge(scanMinecraft());
    merge(scanStandalone());
    // TODO: merge scanOwnedSteam() (requires Steam provider)

    // Filter out proton, sort
    all.erase(
        std::remove_if(all.begin(), all.end(), [](const Game& g) {
            return g.name().toLower().contains("proton");
        }),
        all.end()
    );

    std::sort(all.begin(), all.end(), [](const Game& a, const Game& b) {
        return metadata::sortKey(a) < metadata::sortKey(b);
    });

    return all;
}
